runtime/compiler/optimizer/InterpreterEmulator.cpp

```diff
@@ -723,28 +1262,77 @@ InterpreterEmulator::debugUnresolvedOrCold(TR_ResolvedMethod *resolvedMethod)
       }
    }
 
+void
+InterpreterEmulator::refineResolvedCalleeForInvokevirtual(TR_ResolvedMethod *&callee, bool &isIndirectCall)
+   {
+   TR_ASSERT_FATAL(_iteratorWithState, "has to be called when the iterator has state!");
+   if (!comp()->getOrCreateKnownObjectTable())
+      return;
+
+   TR::RecognizedMethod rm = callee->getRecognizedMethod();
+   switch (rm)
+      {
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+      case TR::java_lang_invoke_MethodHandle_invokeBasic:
+         {
+         int argNum = callee->numberOfExplicitParameters();
+         TR::KnownObjectTable::Index receiverIndex = topn(argNum)->getKnownObjectIndex();
+         TR_J9VMBase* fej9 = comp()->fej9();
+         auto targetMethod = fej9->targetMethodFromMethodHandle(comp(), receiverIndex);
+         if (!targetMethod) return;
+
+         isIndirectCall = false;
+         callee = fej9->createResolvedMethod(comp()->trMemory(), targetMethod, callee->owningMethod());
+         heuristicTrace(tracer(), "Refine invokeBasic to %s\n", callee->signature(trMemory(), stackAlloc));
+         return;
+         }
+#endif //J9VM_OPT_OPENJDK_METHODHANDLE
+      default:
+         return;
+      }
+   }
+
 void
 InterpreterEmulator::visitInvokevirtual()
    {
    int32_t cpIndex = next2Bytes();
    auto calleeMethod = (TR_ResolvedJ9Method*)_calltarget->_calleeMethod;
    bool isUnresolvedInCP;
-   TR_ResolvedMethod * resolvedMethod = calleeMethod->getResolvedPossiblyPrivateVirtualMethod(comp(), cpIndex, true, &isUnresolvedInCP);
+   // Calls in thunk archetype won't be executed by inliner, so they may appear as unresolved
+   bool ignoreRtResolve = _callerIsThunkArchetype;
+   TR_ResolvedMethod * resolvedMethod = calleeMethod->getResolvedPossiblyPrivateVirtualMethod(comp(), cpIndex, ignoreRtResolve, &isUnresolvedInCP);
    Operand *result = NULL;
    if (isCurrentCallUnresolvedOrCold(resolvedMethod, isUnresolvedInCP))
       {
       debugUnresolvedOrCold(resolvedMethod);
       }
    else if (resolvedMethod)
       {
+      bool isIndirectCall = !resolvedMethod->isFinal() && !resolvedMethod->isPrivate();
+      if (_iteratorWithState)
+         refineResolvedCalleeForInvokevirtual(resolvedMethod, isIndirectCall);
+
+      // Customization logic is not needed in customized thunk or in inlining
+      // with known MethodHandle object
+      // Since branch folding is disabled and we're ignoring the coldness info
+      // in thunk archetype, calls to the following method will be added to the
+      // call site list and take up some inlining budget, causing less methods
+      // to be inlined. Don't create call site for them
+      //
+      switch (resolvedMethod->getRecognizedMethod())
```

Comment:
@vijaysun-omr Calls to the following methods are inside the blocks that are folded by branch folding, so they won't be added to the inlining list. Since branch folding is disabled, the following code is to make sure we don't add them to inlining candidate list to save inlining budget for other methods.